ext/coverage/coverage.c

```diff
@@ -99,7 +99,11 @@ rb_coverage_set_result(VALUE klass, VALUE ncoverages)
     if (!RTEST(coverages)) {
 	rb_raise(rb_eRuntimeError, "coverage measurement is not enabled");
     }
-    st_foreach(RHASH_TBL(ncoverages), coverage_merge_result_i, coverages);
+    if (!RTEST(rb_check_hash_type(ncoverages)) {
```

Comment:
`rb_check_hash_type` returns a `Hash` which may be converted and different from the original object.
You have to save it.
Or, just `Check_Type(ncoverages, T_HASH)` may be enough.